tests/test_PixelFlags.py

```diff
@@ -103,28 +108,50 @@ def testOffimageNonfinite(self):
+                                      np.array([False, True, True, True]))
+        np.testing.assert_array_equal(catalog["base_PixelFlags_flag_edgeCenterAll"],
+                                      np.array([False, True, True, True]))
 
     def testOffimage(self):
         """Test that sources at the boundary of the image get flag_offimage
-        and flag_edge set appropriately.
+        and flag_edge[Center[All]] set appropriately.
         """
         # These four will be explicitly set to values at the boundary; we
         # cannot add sources off the image in TestDataset.
         self.dataset.addSource(100000, lsst.geom.Point2D(20, 20))
         self.dataset.addSource(100000, lsst.geom.Point2D(20, 100))
-        self.dataset.addSource(100000, lsst.geom.Point2D(80, 100))
+        self.dataset.addSource(100000, lsst.geom.Point2D(40, 100))
         self.dataset.addSource(100000, lsst.geom.Point2D(80, 30))
+        # Half on the edge; no offimage or edgeCenterAll, but edge and edgeCenter.
+        self.dataset.addSource(100000, lsst.geom.Point2D(115, 30))
+        # Fully on the edge; no offimage but all edge flags.
+        self.dataset.addSource(100000, lsst.geom.Point2D(10, 127))
         task = self.makeSingleFrameMeasurementTask("base_PixelFlags")
         exposure, catalog = self.dataset.realize(10.0, task.schema, randomSeed=0)
+        # All of these should get edgeCenter and edgeCenterAll.
         catalog[1]["slot_Centroid_x"] = -20.5  # on image
         catalog[2]["slot_Centroid_x"] = -20.51  # off image
         catalog[3]["slot_Centroid_y"] = 129.4  # on image
         catalog[4]["slot_Centroid_y"] = 129.50  # off image
+        # Mask 4 pixels at the x-edges
+        edgeBit = exposure.mask.getPlaneBitMask("EDGE")
+        exposure.mask.array[:4, :] |= edgeBit
+        exposure.mask.array[-4:, ] |= edgeBit
+        # Mask 4 pixels at the y-edges
+        nodataBit = exposure.mask.getPlaneBitMask("NO_DATA")
+        exposure.mask.array[:, :4] |= nodataBit
+        exposure.mask.array[:, -4:] |= nodataBit
         task.run(catalog, exposure)
+
         np.testing.assert_array_equal(catalog["base_PixelFlags_flag_offimage"],
-                                      np.array([False, False, True, False, True]))
+                                      np.array([False, False, True, False, True, False, False]))
         np.testing.assert_array_equal(catalog["base_PixelFlags_flag_edge"],
-                                      np.array([False, False, True, False, True]))
+                                      np.array([False, False, True, False, True, True, True]))
+        np.testing.assert_array_equal(catalog["base_PixelFlags_flag_edgeCenter"],
+                                      np.array([False, True, True, True, True, True, True]))
+        np.testing.assert_array_equal(catalog["base_PixelFlags_flag_edgeCenterAll"],
+                                      np.array([False, True, True, True, True, False, True]))
```

Comment:
This style of testing is very hard to validate (which index is which source? 😵‍💫), but I suppose that's out of scope for this issue.

Reply:
Yeah, I know. I am responsible for adding these tests myself, so I can't blame history on this part. It was useful to have them all in one test so I could look at them on the image with afw.display to check that I'd put them where I thought I had, and then to look at the arrays being tested to see that they covered the cases I wanted.

Suggestions on how to clean it up? I could pull each source into its own test, but then it's a bit harder to reason about having caught an appropriate number of cases.

Reply:
Yeah, a separate method for each would probably be best. The "loop over examples" idiom wouldn't really work here, especially if you need explanatory comments for each source. 🙂

Reply:
I separated the edge from the offImage tests, so at least there are many fewer bools in the arrays. I think because they're grouped together and only 2 or 4 entries to check, this is clear enough now. Thoughts?